File: src/ros2_bridge/mappers/trajectory_msgs/joint_trajectory_point.rs
//! Typed mapper for `trajectory_msgs/msg/JointTrajectoryPoint`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn joint_trajectory_point_to_bus(
    msg: ros_env::trajectory_msgs::msg::JointTrajectoryPoint,
) -> crate::trajectory_msgs::msg::v1::JointTrajectoryPoint {
    crate::trajectory_msgs::msg::v1::JointTrajectoryPoint {
        positions: crate::ros2_bridge::mappers::convert::f64_seq(msg.positions),
        velocities: crate::ros2_bridge::mappers::convert::f64_seq(msg.velocities),
        accelerations: crate::ros2_bridge::mappers::convert::f64_seq(msg.accelerations),
        effort: crate::ros2_bridge::mappers::convert::f64_seq(msg.effort),
        time_from_start: Some(
            crate::ros2_bridge::mappers::builtin_interfaces::duration::duration_to_bus(
                msg.time_from_start,
            ),
        ),
    }
}

pub(crate) fn joint_trajectory_point_to_ros(
    bus: crate::trajectory_msgs::msg::v1::JointTrajectoryPoint,
) -> ros_env::trajectory_msgs::msg::JointTrajectoryPoint {
    ros_env::trajectory_msgs::msg::JointTrajectoryPoint {
        positions: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.positions),
        velocities: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.velocities),
        accelerations: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(
            bus.accelerations,
        ),
        effort: crate::ros2_bridge::mappers::convert::FromF64Seq::from_f64_seq(bus.effort),
        time_from_start: crate::ros2_bridge::mappers::builtin_interfaces::duration::duration_to_ros(
            bus.time_from_start.unwrap_or_default(),
        ),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct TrajectoryMsgsJointTrajectoryPointMapper;

impl TypedTopicMapper for TrajectoryMsgsJointTrajectoryPointMapper {
    type Ros = ros_env::trajectory_msgs::msg::JointTrajectoryPoint;
    type Bus = crate::trajectory_msgs::msg::v1::JointTrajectoryPoint;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(joint_trajectory_point_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(joint_trajectory_point_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/trajectory_msgs/joint_trajectory_point.py
"""Generated mapper for `trajectory_msgs/msg/JointTrajectoryPoint`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.builtin_interfaces.duration import duration_to_bus, duration_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.trajectory_msgs.msg.v1 import JointTrajectoryPoint as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def joint_trajectory_point_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.positions.extend(list(msg.positions))
    bus.velocities.extend(list(msg.velocities))
    bus.accelerations.extend(list(msg.accelerations))
    bus.effort.extend(list(msg.effort))
    bus.time_from_start.CopyFrom(duration_to_bus(msg.time_from_start))
    return bus


def joint_trajectory_point_to_ros(bus):
    from trajectory_msgs.msg import JointTrajectoryPoint as RosMsg

    out = RosMsg()
    out.positions = list(bus.positions)
    out.velocities = list(bus.velocities)
    out.accelerations = list(bus.accelerations)
    out.effort = list(bus.effort)
    out.time_from_start = duration_to_ros(bus.time_from_start)
    return out


class TrajectoryMsgsJointTrajectoryPointMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from trajectory_msgs.msg import JointTrajectoryPoint as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return joint_trajectory_point_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return joint_trajectory_point_to_ros(bus)
